src/uct/cuda/cuda_copy/cuda_copy_iface.c

```diff
@@ -469,15 +433,62 @@ static uct_iface_internal_ops_t uct_cuda_copy_iface_internal_ops = {
     .ep_is_connected       = uct_base_ep_is_connected
 };
 
+ucs_status_t uct_cuda_copy_ctx_rsc_create(uct_cuda_copy_iface_t *iface,
+                                          CUcontext ctx,
+                                          unsigned long long ctx_id,
+                                          uct_cuda_copy_ctx_rsc_t **ctx_rsc_p)
+{
+    ucs_kh_put_t ret;
+    khiter_t iter;
+    uct_cuda_copy_ctx_rsc_t *ctx_rsc;
+    ucs_status_t status;
+    ucs_mpool_params_t mp_params;
+    ucs_memory_type_t src, dst;
+
+    iter = kh_put(cuda_copy_ctx_rscs, &iface->ctx_rscs, ctx_id, &ret);
+    if (ret == UCS_KH_PUT_FAILED) {
+        ucs_error("cannot allocate hash entry");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ucs_assertv_always(ret != UCS_KH_PUT_KEY_PRESENT,
+                       "the key %llu has already been added", ctx_id);
+
+    ctx_rsc = &kh_value(&iface->ctx_rscs, iter);
+
+    ucs_mpool_params_reset(&mp_params);
+    mp_params.elem_size       = sizeof(uct_cuda_copy_event_desc_t);
+    mp_params.elems_per_chunk = 128;
+    mp_params.max_elems       = iface->config.max_cuda_events;
+    mp_params.ops             = &uct_cuda_copy_event_desc_mpool_ops;
+    mp_params.name            = "cuda_copy_event_descriptors";
+
+    status = ucs_mpool_init(&mp_params, &ctx_rsc->cuda_event_desc);
+    if (status != UCS_OK) {
+        kh_del(cuda_copy_ctx_rscs, &iface->ctx_rscs, iter);
+        return status;
+    }
+
+    ucs_memory_type_for_each(src) {
+        ucs_memory_type_for_each(dst) {
+            ctx_rsc->queue_desc[src][dst].stream = NULL;
+            ucs_queue_head_init(&ctx_rsc->queue_desc[src][dst].event_queue);
```

Comment:
i think this is a problem to place event_queue inside the hash table value because when the hash size changes the back pointers become invalid